File: benchrec/real.py
"""Client C: a real bank's cash desk (BenchRec), run through the same idea as the simulated clients.

    uv run python -m benchrec.real

1. Deterministic matcher, thresholds already tuned on TRAIN (runs/benchrec_results.json).
2. Playbook of conventions induced from the analysts' own resolutions in TRAIN (benchrec/induce.py).
3. One pass over EVAL with everything frozen. Only `grade()` below reads solution.csv.

Writes runs/real_results.json.
"""

import csv
import json
import random
import subprocess
import time
from collections import Counter, defaultdict
from datetime import date, datetime, timezone
from pathlib import Path

from . import data
from .induce import FLOOR, analyst_groups, apply, induce, is_right, linked
from .matcher import Blocks, Thresholds, match
from .scoring import score

ROOT = Path(__file__).resolve().parent.parent
RESULTS = ROOT / "runs" / "real_results.json"
MATCHER_RESULTS = ROOT / "runs" / "benchrec_results.json"
N_EXAMPLES = 36
SEED = 7


def operators(path=data.TRAIN):
    """matchId -> (rule, operator), and the operator profile. The rule sits on one row of each group."""
    rule_of, by_of = {}, {}
    with open(path, newline="", encoding="utf-8-sig") as fh:
        for r in csv.DictReader(fh):
            by_of[r["matchId"]] = r["matchedBy"]
            if r["matchRule"]:
                rule_of[r["matchId"]] = r["matchRule"]
    return rule_of, by_of


def _q(xs, p):
    xs = sorted(xs)
    return xs[min(len(xs) - 1, int(p * len(xs)))] if xs else None


def operator_profile(ledger, bank, rule_of, by_of):
    groups = defaultdict(lambda: ([], []))
    for a in ledger:
        groups[a.match_id][0].append(a)
    for b in bank:
        groups[b.match_id][1].append(b)
    manual_ops = {by_of[g] for g, r in rule_of.items() if r == "MANUAL"}
    prof = defaultdict(lambda: defaultdict(list))
    for g, (As, Bs) in groups.items():
        p = prof[by_of[g]]
        p["groups"].append(1)
        p["manual"].append(rule_of.get(g) == "MANUAL")
        p["one_to_one"].append(len(As) == 1 and len(Bs) == 1)
        p["no_ledger_side"].append(not As)
        if len(As) == 1 and len(Bs) == 1:
            p["amount_exact"].append(As[0].cents == Bs[0].cents)
            p["same_day"].append(As[0].day == Bs[0].day)
            if As[0].cents != Bs[0].cents:
                p["diff_cents"].append(abs(As[0].cents - Bs[0].cents))
    out = []
    for op, p in sorted(prof.items(), key=lambda kv: -len(kv[1]["groups"])):
        share = lambda k: (sum(p[k]) / len(p[k])) if p[k] else None  # noqa: E731
        out.append({
            "operator": op,
            "kind": "analyst" if op in manual_ops else "matching engine (inferred: never closes anything manually)",
            "groups": len(p["groups"]),
            "share_manual": share("manual"),
            "share_one_to_one": share("one_to_one"),
            "share_groups_with_no_ledger_side": share("no_ledger_side"),
            "one_to_one_groups": len(p["amount_exact"]),
            "one_to_one_with_a_difference": len(p["diff_cents"]),
            "one_to_one_amount_exact": share("amount_exact"),
            "one_to_one_same_value_date": share("same_day"),
            "median_accepted_difference_cents": _q(p["diff_cents"], 0.5),
            "p90_accepted_difference_cents": _q(p["diff_cents"], 0.9),
        })
    analysts = {g for g, op in by_of.items() if op in manual_ops}
    return out, analysts


def grade(matches, cleared, truth):
    """The only place the answer key is used."""
    base = {m.b_id: m.allocations for m in matches}
    s0 = score(base, truth)
    preds = dict(base)
    per = defaultdict(lambda: {"bank_lines": 0, "right": 0})
    no_ledger = {"bank_lines": 0, "right": 0}
    for r in cleared:
        ok = is_right(r, truth)
        bucket = no_ledger if r.verdict == "no_ledger" else per[r.convention]
        bucket["bank_lines"] += len(r.b_ids)
        bucket["right"] += len(r.b_ids) * ok
        if r.verdict == "no_ledger":
            per[r.convention]["bank_lines"] += len(r.b_ids)
            per[r.convention]["right"] += len(r.b_ids) * ok
        else:
            for b in r.b_ids:
                preds[b] = r.allocations
    s1 = score(preds, truth)
    for v in per.values():
        v["precision"] = v["right"] / v["bank_lines"] if v["bank_lines"] else None
    return s0, s1, dict(per), no_ledger


def _line(rec):
    return {"id": rec.id, "side": "ledger" if rec.side == "A" else "bank", "amount": rec.cents / 100,
            "value_date": date.fromordinal(rec.day).isoformat(), "dc": rec.dc,
            "references": rec.refs.strip()[:120], "attributes": " ".join(rec.attrs.split())[:160]}


def examples(cleared, asked, rest, ledger, bank, playbook, truth):
    a_by, b_by = {a.id: a for a in ledger}, {b.id: b for b in bank}
    conv = {c.id: c for c in playbook}
    rng = random.Random(SEED)
    picks = []
    by_conv = defaultdict(list)
    for r in cleared:
        by_conv[("cleared", r.convention)].append(r)
    for r in asked:
        by_conv[("escalated", r.convention)].append(r)
    for r in rest:
        by_conv[("escalated", r.reason)].append(r)
    per = max(3, N_EXAMPLES // max(1, len(by_conv)))
    for key in sorted(by_conv):
        rs = by_conv[key]
        wrong = [r for r in rs if r.verdict != "escalated" and not is_right(r, truth)]
        chosen = rng.sample(rs, min(per, len(rs)))
        if wrong and not any(r in wrong for r in chosen):
            chosen[-1] = rng.choice(wrong)  # always show a miss where there is one
        picks += chosen
    out = []
    a_amounts = defaultdict(list)
    for a in ledger:
        a_amounts[a.cents].append(a)
    for r in picks[: N_EXAMPLES + 12]:
        c = conv.get(r.convention)
        cands = [a_by[i] for i in r.a_ids]
        if not cands and r.verdict == "escalated":
            b0 = b_by[r.b_ids[0]]
            cands = sorted(a_amounts.get(b0.cents, []), key=lambda a: abs(a.day - b0.day))[:3]
        if r.verdict == "escalated":
            right = None
            would = _suggestion_right(r, truth) if r.convention else None
        else:
            right, would = is_right(r, truth), None
        out.append({
            "verdict": r.verdict,
            "convention": r.convention or None,
            "convention_title": c.title if c else None,
            "reason": r.reason,
            "measured": f"{r.param} {c.unit}" if c else None,
            "band": {"applies_to": c.applies_to, "asks_to": c.asks_to, "unit": c.unit} if c else None,
            "precedents": ([c.applies_precedent] + c.precedents[:2]) if c else [],
            "bank_lines": [_line(b_by[i]) for i in r.b_ids],
            "ledger_lines": [_line(a) for a in cands],
            "reference_found_in_bank_text": [linked(a, b_by[r.b_ids[0]]) for a in cands],
            "right_per_scorer": right,
            "suggestion_would_have_been_right": would,
            "bank_line_has_a_ledger_match_in_key": bool(truth.get(r.b_ids[0])),
        })
    return out


def _suggestion_right(r, truth):
    """An escalated suggestion, graded as if it had been applied. An offset suggests 'no ledger side'."""
    if r.convention == "bank_side_offset":
        return all(not truth.get(b) for b in r.b_ids)
    return bool(r.allocations) and all(truth.get(b) and r.allocations <= truth[b] for b in r.b_ids)


def commit():
    try:
        return subprocess.run(["git", "rev-parse", "--short", "HEAD"], cwd=ROOT, capture_output=True,
                              text=True, check=True).stdout.strip()
    except Exception:
        return None


def main():
    t0 = time.time()
    th = Thresholds(**json.loads(MATCHER_RESULTS.read_text())["thresholds"])
    train_a, train_b = data.load_records(data.TRAIN)
    rule_of, by_of = operators()
    profile, analysts = operator_profile(train_a, train_b, rule_of, by_of)
    train_matches = match(Blocks(train_a, train_b), th)
    playbook = induce(train_a, train_b, train_matches, analysts)
    print("INDUCED ON TRAIN")
    for c in playbook:
        bt = c.backtest.get("at_applies_to") or c.backtest.get("at_widest") or {}
        print(f"  {c.id:<20} support {c.support:>5}  applies<= {c.applies_to}  asks<= {c.asks_to}  "
              f"executes {c.executes}  backtest {bt.get('agree')}/{bt.get('bank_lines')}")

    # frozen from here on
    eval_a, eval_b = data.load_records(data.EVAL)
    eval_matches = match(Blocks(eval_a, eval_b), th)
    cleared, asked, rest = apply(eval_a, eval_b, eval_matches, playbook)

    truth = data.load_solution()
    s0, s1, per, no_ledger = grade(eval_matches, cleared, truth)
    base = score(data.load_baseline(), truth)
    n = s0["n_bank_lines"]
    n_residue = n - s0["matched"]
    lines_cleared = sum(len(r.b_ids) for r in cleared)
    lines_matched_by_conv = sum(len(r.b_ids) for r in cleared if r.verdict != "no_ledger")
    right_conv = sum(v["right"] for k, v in per.items() if k != "bank_side_offset")
    asked_lines = sum(len(r.b_ids) for r in asked)
    reasons = Counter(r.reason for r in rest)
    sugg_right = sum(len(r.b_ids) for r in asked if _suggestion_right(r, truth))

    def tier(s):
        return {"matched": s["matched"], "correct_pair": s["correct_pair_level"],
                "precision_pair": s["precision_pair_level"], "match_rate_pair": s["match_rate_pair_level"],
                "precision_strict": s["precision"], "match_rate_strict": s["match_rate"],
                "left_for_review": s["left_for_review"], "wrong_pair": s["matched"] - s["correct_pair_level"]}

    days = [b.day for b in eval_b]
    tdays = [b.day for b in train_b]
    results = {
        "client": "Client C: a real bank's cash desk",
        "dataset": "BenchRec cash v1.0 (Kaggle, CC BY 4.0): real, anonymised bank-to-ledger lines labelled by the bank's analysts",
        "generated_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "code_commit": commit(),
        "protocol": [
            "Matcher thresholds tuned on the train split only.",
            "Conventions induced from the analysts' resolutions in the train split only; bands chosen by back-test on train.",
            "The eval split is the dataset authors' own held-out file. Its value dates overlap the end of the train period, so it is held out by line, not by calendar.",
            "One pass over the eval split with everything frozen. The answer key (solution.csv) is opened only by the grader.",
            f"A convention executes only if it agrees with the analysts on at least {FLOOR:.0%} of the train lines it touches.",
            "Correct means pair level: every ledger allocation we attach belongs to that bank line's labelled match. Strict (exact set) is reported beside it.",
        ],
        "train": {"bank_lines": len(train_b), "ledger_lines": len(train_a),
                  "from": date.fromordinal(min(tdays)).isoformat(), "to": date.fromordinal(max(tdays)).isoformat(),
                  "analyst_closed_groups": len(analysts)},
        "eval": {"bank_lines": n, "matchable": s0["n_matchable"], "ledger_lines": len(eval_a),
                 "from": date.fromordinal(min(days)).isoformat(), "to": date.fromordinal(max(days)).isoformat()},
        "tiers": {
            "shipped_baseline": tier(base),
            "matcher": tier(s0),
            "matcher_plus_conventions": tier(s1),
        },
        "residue": {
            "bank_lines_left_by_matcher": n_residue,
            "cleared_by_conventions": lines_cleared,
            "matched_to_ledger_by_conventions": lines_matched_by_conv,
            "right_pair": right_conv,
            "precision_pair": right_conv / lines_matched_by_conv if lines_matched_by_conv else None,
            "declared_no_ledger_side": no_ledger,
            "share_of_residue_cleared": lines_cleared / n_residue if n_residue else None,
            "escalated_with_a_suggestion": asked_lines,
            "suggestions_that_were_right": sugg_right,
            "escalated_other": dict(reasons),
            "still_for_a_person": asked_lines + len(rest),
            "model_calls": 0,
            "model_cost_usd": 0.0,
        },
        "by_convention_on_eval": per,
        "playbook": [c.as_dict() for c in playbook],
        "operators": profile,
        "account_contrast": {
            "possible": False,
            "finding": "The dataset holds one account (ACC#00001) in one currency (USD), so it cannot show two clients "
                       "or two accounts disagreeing. What it does show is who resolves what: the engine closes "
                       "exact amounts on the same value date (all but a handful of its matches), and the people close everything else under "
                       "conventions the engine does not have.",
        },
        "examples": examples(cleared, asked, rest, eval_a, eval_b, playbook, truth),
        "runtime_seconds": round(time.time() - t0, 1),
        "notes": [
            "No model was called anywhere on this page. Every number is from deterministic code and costs $0 to run.",
            "The scoring script used by the dataset's authors was not available; the grader is our own reading of solution.csv, "
            "and the shipped baseline is graded with the same function.",
            "A bank-side offset is graded right when the key lists no ledger match for either line. It is reported apart "
            "from precision and match rate, which only count lines matched to the ledger.",
            "Operator names are the dataset's own anonymised labels. Which one is the engine is our inference.",
        ],
    }
    RESULTS.parent.mkdir(exist_ok=True)
    RESULTS.write_text(json.dumps(results, indent=1))

    print("\nEVAL (frozen)")
    for k, t in results["tiers"].items():
        print(f"  {k:<26} matched {t['matched']:>6}  precision {t['precision_pair']:.4%}  match rate {t['match_rate_pair']:.4%}"
              f"  strict {t['precision_strict']:.4%}/{t['match_rate_strict']:.4%}  wrong {t['wrong_pair']}")
    print(f"  residue {n_residue}: conventions cleared {lines_cleared} ({right_conv}/{lines_matched_by_conv} right, "
          f"no-ledger {no_ledger}), asked {asked_lines} (suggestion right {sugg_right}), other {dict(reasons)}")
    for k, v in per.items():
        print(f"    {k:<20} {v}")
    print(f"  wrote {RESULTS} in {results['runtime_seconds']}s")


if __name__ == "__main__":
    main()

File: ui/js/real.js
/* Real ledger: GET /api/real/results, /api/real/playbook, /api/real/examples.
   BenchRec, a real bank's anonymised cash ledger. No model is called anywhere on this page. */
(async function () {
  const { get, usd, esc, day } = SO;
  const view = document.getElementById("view"), summary = document.getElementById("summary");
  const pct = (x, d) => (x == null ? "n/a" : (x * 100).toFixed(d ?? 2) + "%");
  const int = (n) => Number(n).toLocaleString("en-US");
  const REASON = {
    no_ledger_candidate: "No ledger entry of that amount exists yet",
    several_same_amount_candidates: "Several ledger entries of that amount, nothing to tell them apart",
    one_same_amount_candidate_outside_every_band: "One candidate, but outside every band the history supports",
  };

  function bars(title, note, rows, max, fmt) {
    return `<div class="chart"><div class="chart-head"><h3>${esc(title)}</h3><span class="faint small">${esc(note)}</span></div>
      ${rows.map((r) => `<div class="bar-row"><span class="bar-label">${esc(r.label)}</span>
        <span class="bar-track"><span class="mk ${r.base ? "base" : ""}" style="width: ${Math.max(0.5, (r.value / max) * 100)}%"></span></span>
        <span class="bar-value num">${esc(fmt(r.value))}</span></div>`).join("")}</div>`;
  }

  // same three zones as the playbook screen; log scale because a band can run from cents to dollars
  function band(c) {
    const money = c.unit === "cents";
    const show = (v) => (money ? usd(v / 100) : `${v} ${c.unit}`);
    const max = Math.max(c.asks_to || 1, c.applies_to || 1) * 1.6;
    const pos = (v) => Math.min(100, (Math.log(1 + v) / Math.log(1 + max)) * 100);
    const lo = c.executes ? pos(c.applies_to) : 0, hi = pos(c.asks_to);
    const end = (at, v, p, word) => `<span class="band-end ${at < 8 ? "at-start" : at > 92 ? "at-end" : ""}" style="left: ${at}%"><b class="num">${esc(show(v))}</b><span class="faint">${word} <span class="cite">${esc(p)}</span></span></span>`;
    return `<div class="band"><div class="band-track">
        <span class="band-seg acts" style="left: 0; width: ${lo}%"></span>
        <span class="band-seg asks" style="left: ${lo}%; width: ${hi - lo}%"></span></div>
      <div class="band-ends">${c.executes ? end(lo, c.applies_to, c.applies_precedent, "precedent") : ""}${end(hi, c.asks_to, c.asks_precedent, "widest seen")}</div>
      <div class="legend"><span><i class="band-key acts"></i>${c.executes ? "Applies on its own, $0.00" : "Automatic execution disabled"}</span><span><i class="band-key asks"></i>Asks a person</span><span><i class="band-key out"></i>Does not apply</span></div></div>`;
  }

  function convention(c, onEval) {
    const bt = c.backtest.at_applies_to || c.backtest.at_widest || {};
    const e = onEval[c.id];
    return `<section class="panel"><div class="panel-head"><h3>${esc(c.title)}</h3>
        <span class="runrow"><span class="state ${c.executes ? "state-book" : "state-proposed"}">${c.executes ? "Executes" : "Held back by its own history"}</span></span></div>
      <div class="panel-body stack"><p style="max-width: 78ch;">${esc(c.statement)}</p>
        ${c.asks_to != null ? band(c) : ""}
        <dl class="kv num">
          <dt>Analyst precedents</dt><dd>${int(c.support)} match groups closed by a person</dd>
          <dt>Back-test on history</dt><dd>${bt.bank_lines ? `agrees with the analysts on ${int(bt.agree)} of ${int(bt.bank_lines)} lines (${pct(bt.agreement, 1)})${c.executes ? "" : ", under the 98% it needs to run"}` : "nothing to test it on"}</dd>
          ${e ? `<dt>On the held-out split</dt><dd>${int(e.right)} of ${int(e.bank_lines)} right (${pct(e.precision, 1)})</dd>` : ""}
          <dt>Cites</dt><dd>${(c.precedents || []).map((p) => `<span class="cite">${esc(p)}</span>`).join(" ")}</dd>
        </dl>${c.note ? `<p class="muted small" style="max-width: 78ch;">${esc(c.note)}</p>` : ""}</div></section>`;
  }

  function line(l, found) {
    return `<tr><td>${l.side === "bank" ? "Bank" : "Ledger"}</td><td class="num" style="text-align: right;">${esc(usd(l.amount))}</td><td class="num">${esc(day(l.value_date))}</td>
      <td class="mono small">${esc(l.references || "")}${found ? ` <span class="state state-book">reference found in bank text</span>` : ""}<div class="faint">${esc(l.attributes || "")}</div></td></tr>`;
  }

  function example(x) {
    const graded = x.verdict === "escalated"
      ? (x.suggestion_would_have_been_right == null ? "" : `<span class="state state-proposed">${x.suggestion_would_have_been_right ? "Suggestion matched the key" : "Suggestion did not match the key"}</span>`)
      : `<span class="state ${x.right_per_scorer ? "state-book" : "state-escalate"}">${x.right_per_scorer ? "Right per the answer key" : "Wrong per the answer key"}</span>`;
    const verdict = x.verdict === "cleared" ? "Cleared at $0.00" : x.verdict === "no_ledger" ? "Closed, no ledger side" : "Sent to a person";
    return `<details class="panel ex"><summary class="panel-head" style="cursor: pointer;"><h3 style="font-size: var(--t-md);">${esc(verdict)} · <span class="num">${esc(usd(x.bank_lines[0].amount))}</span> <span class="faint">${esc(x.convention_title || REASON[x.reason] || x.reason)}</span></h3><span class="runrow">${graded}</span></summary>
      <div class="panel-body stack">
        <div class="table-wrap"><table><thead><tr><th>Side</th><th style="text-align: right;">Amount</th><th>Value date</th><th>References</th></tr></thead><tbody>
          ${x.bank_lines.map((l) => line(l)).join("")}${x.ledger_lines.map((l, i) => line(l, x.reference_found_in_bank_text[i])).join("")}</tbody></table></div>
        <dl class="kv num"><dt>Why</dt><dd>${esc(REASON[x.reason] || x.reason)}</dd>
          ${x.measured ? `<dt>Measured</dt><dd>${esc(x.measured)}${x.band && x.band.applies_to != null ? `, band applies to ${esc(String(x.band.applies_to))} and asks to ${esc(String(x.band.asks_to))} ${esc(x.band.unit)}` : ""}</dd>` : ""}
          ${(x.precedents || []).filter(Boolean).length ? `<dt>Precedents</dt><dd>${x.precedents.filter(Boolean).map((p) => `<span class="cite">${esc(p)}</span>`).join(" ")}</dd>` : ""}
          ${x.ledger_lines.length === 0 ? `<dt>Ledger</dt><dd>No candidate entry found</dd>` : ""}</dl></div></details>`;
  }

  try {
    const [r, pb, ex] = await Promise.all([get("/api/real/results"), get("/api/real/playbook"), get("/api/real/examples")]);
    const t = r.tiers, base = t.shipped_baseline, m = t.matcher, mc = t.matcher_plus_conventions, z = r.residue;
    summary.innerHTML = `<span class="state state-book">Real data</span><span><b>${int(r.eval.bank_lines)}</b> bank lines, ${esc(day(r.eval.from))} to ${esc(day(r.eval.to))}. Learned from <b>${int(r.train.bank_lines)}</b> training lines and <b>${int(r.train.analyst_closed_groups)}</b> matches the bank's own analysts closed by hand. Run once with everything frozen. <b>0</b> model calls, <b>$0.00</b>.</span>`;
    const tierRows = (key) => [{ label: "Reference matcher shipped with the data", value: base[key], base: true }, { label: "Our matcher", value: m[key] }, { label: "Matcher plus induced conventions", value: mc[key] }];
    const other = Object.entries(z.escalated_other).sort((a, b) => b[1] - a[1]);
    const head = `<section class="panel"><div class="panel-head"><h3>BenchRec cash v1.0 · held-out evaluation</h3><span class="runrow"><span class="faint small">Code ${esc(r.code_commit || "uncommitted")} · ${esc(r.generated_at.slice(0, 16).replace("T", " "))} UTC</span></span></div>
      <div class="panel-body stack">
        <h2 class="ask-q" style="max-width: 44ch;"><span class="num">${pct(mc.match_rate_pair, 1)}</span> of matchable bank lines matched at <span class="num">${pct(mc.precision_pair, 2)}</span> pair-level precision. Zero LLM calls.</h2>
        <p style="max-width: 85ch;">Tested on <b>${int(r.eval.bank_lines)} real, anonymised bank transactions</b> from BenchRec, a bank-to-ledger reconciliation benchmark. ${int(r.eval.matchable)} have a labelled ledger match. Our benchmark matcher plus learned conventions correctly matched <b>${int(mc.correct_pair)}</b> of those lines and made <b>${int(mc.wrong_pair)} incorrect matches</b> among ${int(mc.matched)} predictions.</p>
        <p class="muted" style="max-width: 85ch;">Thresholds and conventions were learned from ${int(r.train.bank_lines)} training bank lines, then frozen for evaluation. No hand-entered company policy rules; no LLM inference on this benchmark. The training and evaluation dates overlap, so this is a held-out set of transactions, not a future-period test.</p>
        <p class="muted" style="max-width: 80ch;">The reference matcher that ships with the dataset reaches ${pct(base.match_rate_pair, 1)} at ${pct(base.precision_pair, 2)}. Correct means every ledger entry we attach belongs to the match the analysts recorded for that bank line. This permits a partial match to a larger group. Requiring the complete, exact set gives ${pct(mc.precision_strict, 2)} precision and ${pct(mc.match_rate_strict, 2)} coverage of matchable lines.</p>
        <p class="small muted" style="max-width: 85ch;">Scored with our interpretation of the supplied labels, using the same scorer for both systems; this is not an official leaderboard result. It evaluates the BenchRec matching implementation, not the full company-upload or LLM review workflow. <a href="https://www.operartis.com/benchrec" target="_blank" rel="noopener">About the benchmark ↗</a> · <a href="#methodology">Scoring and limitations ↓</a></p>
        <div class="charts">
          ${bars("Bank lines matched correctly", "Share of matchable lines. Scale 0 to 100%", tierRows("match_rate_pair"), 1, (v) => pct(v, 1))}
          ${bars("Wrong matches", "Count, of the matches made. Lower is better", tierRows("wrong_pair"), Math.max(base.wrong_pair, 1), (v) => int(v))}
          ${bars("Precision", "Of the matches made. Scale 0 to 100%", tierRows("precision_pair"), 1, (v) => pct(v, 2))}
          ${bars("Exact-set precision", "Complete labelled group, of predictions made", tierRows("precision_strict"), 1, (v) => pct(v, 2))}
          ${bars("Exact-set coverage", "Share of matchable lines, predicted set equals the labelled set", tierRows("match_rate_strict"), 1, (v) => pct(v, 1))}
        </div></div></section>`;
    const residue = `<section class="panel"><div class="panel-head"><h3>What the matcher left: ${int(z.bank_lines_left_by_matcher)} lines</h3><span class="runrow"><span class="faint small">${pct(m.left_for_review, 1)} of volume</span></span></div>
      <div class="panel-body stack">
        <div class="chart"><div class="bar-row"><span class="bar-label">Where they went</span><span class="bar-track stack-track">
          <span class="mk t1" style="width: ${(z.cleared_by_conventions / z.bank_lines_left_by_matcher) * 100}%" title="Cleared by a convention"></span>
          <span class="mk t2" style="width: ${(z.escalated_with_a_suggestion / z.bank_lines_left_by_matcher) * 100}%" title="To a person, with a suggestion"></span>
          <span class="mk t3" style="width: ${((z.still_for_a_person - z.escalated_with_a_suggestion) / z.bank_lines_left_by_matcher) * 100}%" title="To a person, with a reason"></span></span>
          <span class="bar-value num">${int(z.bank_lines_left_by_matcher)}</span></div>
          <div class="legend"><span><i class="mk t1"></i>Cleared by a convention, $0.00</span><span><i class="mk t2"></i>To a person with a suggestion</span><span><i class="mk t3"></i>To a person with a reason</span></div></div>
        <dl class="kv num">
          <dt>Cleared by induced conventions</dt><dd>${int(z.cleared_by_conventions)} lines, ${int(z.right_pair)} right (${pct(z.precision_pair, 1)})</dd>
          <dt>Sent to a person with a suggestion</dt><dd>${int(z.escalated_with_a_suggestion)} lines. The suggestion matched the key on ${int(z.suggestions_that_were_right)} (${pct(z.suggestions_that_were_right / z.escalated_with_a_suggestion, 0)}), which is why those conventions ask instead of act</dd>
          ${other.map(([k, n]) => `<dt>${esc(REASON[k] || k)}</dt><dd>${int(n)} lines</dd>`).join("")}
        </dl></div></section>`;
    const ops = `<section class="panel"><div class="panel-head"><h3>Who resolves what at this bank</h3><span class="state state-proposed">Measured on history</span></div>
      <div class="panel-body stack"><p style="max-width: 80ch;">${esc(r.account_contrast.finding)}</p>
        <div class="table-wrap"><table><thead><tr><th>Operator</th><th>Kind</th><th style="text-align: right;">Match groups</th><th style="text-align: right;">Closed by hand</th><th style="text-align: right;">One to one</th><th style="text-align: right;">Exact amount</th><th style="text-align: right;">Same value date</th><th style="text-align: right;">Median difference accepted</th></tr></thead><tbody>
        ${r.operators.map((o) => `<tr><td class="mono">${esc(o.operator)}</td><td>${esc(o.kind.split(" (")[0])}</td><td class="num" style="text-align: right;">${int(o.groups)}</td><td class="num" style="text-align: right;">${pct(o.share_manual, 0)}</td><td class="num" style="text-align: right;">${pct(o.share_one_to_one, 0)}</td><td class="num" style="text-align: right;">${pct(o.one_to_one_amount_exact, 0)}</td><td class="num" style="text-align: right;">${pct(o.one_to_one_same_value_date, 0)}</td><td class="num" style="text-align: right;">${o.one_to_one_with_a_difference < 30 ? `too few to say (${int(o.one_to_one_with_a_difference)})` : usd(o.median_accepted_difference_cents / 100)}</td></tr>`).join("")}
        </tbody></table></div><p class="faint small">Operator names are the dataset's own anonymised labels. Which one is the engine is our inference: it never closes anything by hand.</p></div></section>`;
    const convs = `<div class="lede" style="margin-top: var(--s-5);"><div class="label">Induced playbook</div><h2>${pb.conventions.filter((c) => c.executes).length} of ${pb.conventions.length} conventions earned the right to run</h2><p class="muted">Each one is read out of how the analysts closed matches the engine could not, and must agree with them on 98% of the history it touches before it may act. The rest become suggestions for a person.</p></div>
      ${pb.conventions.map((c) => convention(c, pb.by_convention_on_eval)).join("")}`;
    const exs = `<div class="lede" style="margin-top: var(--s-5);"><div class="label">Evidence</div><h2>${ex.examples.length} lines from the held-out split</h2><p class="muted">A fixed random draw, one miss included wherever a convention has one. Open a line to see both sides and the precedent behind the verdict.</p></div>
      ${ex.examples.map(example).join("")}`;
    const notes = `<section class="panel" id="methodology"><div class="panel-head"><h3>How this was measured</h3></div><div class="panel-body"><ul class="muted small" style="max-width: 90ch; display: grid; gap: 6px; padding-left: 1.1em;">${r.protocol.concat(r.notes).map((n) => `<li>${esc(n)}</li>`).join("")}<li>${esc(r.dataset)}.</li></ul></div></section>`;
    view.innerHTML = `<div class="stack">${head}${residue}${ops}${r.summary_only ? `<section class="panel"><div class="panel-body"><h3>Saved benchmark summary</h3><p>This checkout includes verified aggregate results from ${esc(r.generated_at.slice(0, 10))}. Detailed conventions and transaction examples require the original benchmark report. No evaluation was rerun to display these results.</p></div></section>` : convs + exs}${notes}</div>`;
  } catch (e) {
    view.innerHTML = `<div class="panel error">${esc(e.message)}</div>`;
  }
})();
